universe/Building.h

```diff
@@ -214,9 +214,8 @@ class BuildingTypeManager {
 
 private:
     BuildingTypeManager();
-    ~BuildingTypeManager();
 
-    std::map<std::string, BuildingType*> m_building_types;
+    std::map<std::string, std::unique_ptr<BuildingType>> m_building_types;
```

Comment:
To me returning...

`raw T*` indicates nothing about ownership. Returning one indicates that smart pointers aren't be used, and careful consideration of whether and when to delete is required.

`shared_ptr` indicates that the pointed-to object can be used as needed and the pointer can be allowed to go out of scope safely, and can always be assumed to point to a valid object if it did so when initially acquired, and that there may be multiple pointers to the object in existence simultaneously.

`weak_ptr` indicates that the pointed-to object may stop existing after the pointer is acquired, so needs to be converted to a `shared_ptr` before use.

`unique_ptr` indicates that the calling code is receiving the only valid pointer and needs to save it if the object is to continue existing / remain valid.

Since for all the things like `BuildingType` in the manager classes, there can be multiple pointers to it at the same time, they should be held in `shared_ptr`. Using a `unique_ptr` appears very wrong, because it will always need to be possible to have at least two pointers to the object exist simultaneously: within the manager, and within code that asks the manager for a pointer to the object by name or id. (Holding it in a `unique_ptr` and then returning a raw pointer seems even worse, though is probably equivalent by necessity.)

So it's fine to return a `unique_ptr` from a generator class, but if the pointed-to object is meant to continue existing within a manager class and have multiple pointers to it existing at the same time, it should be immediately assigned to a `shared_ptr` to be kept within the manager.